torch/csrc/Module.cpp

```diff
@@ -570,6 +573,118 @@ PyObject *THPModule_fromDLPack(PyObject *_unused, PyObject *data)
   return torch::createPyObject(atensor);
 }
 
+// In cases like data loader, if a worker process die due to bus error/segfault
+// or just hang, the main process, if implemented with multiprocessing.queue.,
+// will hang waiting for data. This is difficult to avoid on PyTorch side as it
+// can be caused by limited shm, other libraries users call in the workers. The
+// following methods is an effort to do our best provide some error message to
+// users when such unfortunate events happen.
+
+// Critical signal handlers should be registered on worker processes before
+// doing work.
+// Python handle is _set_worker_signal_handlers().
+#define SIGNAL_HANDLER(SIGNAL, HANDLER_NAME, ERROR_MSG)                       \
+static void HANDLER_NAME(int sig)                                             \
+{                                                                             \
+    write(fileno(stderr), ERROR_MSG, strlen(ERROR_MSG));                      \
+    _exit(EXIT_FAILURE);                                                      \
+}
+
+SIGNAL_HANDLER(SIGBUS, handler_SIGBUS, "ERROR: Unexpected bus error encountered in worker. "
+  "This might be caused by insufficient shared memory (shm).\n");
+SIGNAL_HANDLER(SIGSEGV, handler_SIGSEGV, "ERROR: Unexpected segmentation fault encountered in worker.\n");
+
+static std::mutex worker_pid_mutex;
+static std::vector<pid_t> worker_pid_vec = {};
+// The following are needed since std::vector is not asynchronous safe.
+static pid_t *worker_pids;
+static size_t num_worker_pids = 0;
+
+PyObject *THPModule_setWorkerSignalHandlers(PyObject *module, PyObject *arg) {
+    signal(SIGBUS, handler_SIGBUS);
+    signal(SIGSEGV, handler_SIGSEGV);
+    Py_RETURN_NONE;
+}
+
+// SIGCHLD hander should be registered on main loader process to catch any
+// worker failing. SIGALRM handler is needed for implementing timeout.
+// Python handles are _set_main_signal_handers() and
+// _remove_main_signal_handers().
+static void handler_SIGCHLD_main(int sig) {
+  int status;
+  pid_t p;
+  pid_t *pid_ptr;
+
+  // The flags and status checks ensure that we are really observing a child
+  // exiting, rather than other cases such as SIGSTOP and SIGCONT.
+  // https://stackoverflow.com/a/40707100
+  while ((p = waitpid(-1, &status, WNOHANG|WUNTRACED|WCONTINUED)) > 0) {
+    if (WIFCONTINUED(status) || WIFSTOPPED(status))
+      continue;
+    if (WIFEXITED(status) != 0 && WEXITSTATUS(status) == 0)
+      continue;
+    // child must have exited with signal/error, check if it is one of the pid
+    // we care about.
+    pid_ptr = worker_pids;
+    for (size_t i = 0; i < num_worker_pids; i++) {
+      if (*pid_ptr == p)
+        _exit(EXIT_FAILURE);
```

Comment:
Does this exit the main-process without an error message?

Reply:
Yes. Printing something will cause the annoying behavior when the main proc errors, it kills children, then this handler gets called and prints "Worker exited unexpectedly in DataLoader", which doesn't make much sense. In such case, the main proc errors, something should already be printed onto stderr.
In case where worker died, we should get error message / segfault / bus error printing from worker. 
So there should always be something on stderr.